core/remote_state_verifier.go

```diff
@@ -69,14 +71,26 @@ func NewVerifyManager(blockchain *BlockChain, peers verifyPeers, allowInsecure b
 		if oldBlock == nil {
 			return nil, fmt.Errorf("block is nil, number: %d", number)
 		}
-		blockHash := oldBlock.Hash()
-		_, err := blockchain.GenerateDiffLayer(blockHash)
-		if err != nil {
-			return nil, err
+
+		// rewind to last non verified block
+		blockchain.SetHead(oldBlock.NumberU64())
+		block = oldBlock
+		break
+	}
+
+	number = block.Number()
+	for i := maxForkHeight; i >= 0; i-- {
+		if new(big.Int).Sub(number, big.NewInt(int64(i))).Cmp(common.Big0) <= 0 {
+			continue
 		}
-		diffLayerCh := make(chan struct{})
-		close(diffLayerCh)
-		blockchain.diffLayerChanCache.Add(blockHash, diffLayerCh)
+		oldBlock := blockchain.GetBlockByNumber(number.Uint64() - uint64(i))
```

Comment:
Use GetBlockByHash not GetBlockByNumber will be better, we should cache parent block

Reply:
fixed